mlir/lib/Transforms/Utils/OneToNTypeConversion.cpp

```diff
@@ -412,4 +413,62 @@ applyPartialOneToNConversion(Operation *op, OneToNTypeConverter &typeConverter,
   return success();
 }
 
+namespace {
+class FunctionOpInterfaceSignatureConversion : public OneToNConversionPattern {
+public:
+  FunctionOpInterfaceSignatureConversion(StringRef functionLikeOpName,
+                                         MLIRContext *ctx,
+                                         TypeConverter &converter)
+      : OneToNConversionPattern(converter, functionLikeOpName, /*benefit=*/1,
+                                ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, OneToNPatternRewriter &rewriter,
+                                const OneToNTypeMapping &operandMapping,
+                                const OneToNTypeMapping &resultMapping,
+                                ValueRange convertedOperands) const override {
+    auto funcOp = cast<FunctionOpInterface>(op);
```

Comment:
Is this *guaranteed* to work? I suppose that the base already tests for `functionLikeOpName`? But what if that op doesn't implement the `FunctionOpInterface`?. Maybe there should an assert here? (I *guess* it should not be a match failure because whoever used the pattern must have thought that the op actually does implement the interface...)

Reply:
The pattern gets an operation name as a constructor arg and only such ops are matched. Using this pattern with ops that do not implement `FunctionOpInterface` is incorrect API usage. `cast` will fail in such a case (I think with a failed assertion internally).

Reply:
You are right! Thanks for helping me see it!